packages/snaps-rpc-methods/src/permitted/setState.test.ts

```diff
@@ -0,0 +1,443 @@
+import { JsonRpcEngine } from '@metamask/json-rpc-engine';
+import { errorCodes } from '@metamask/rpc-errors';
+import type { SetStateResult } from '@metamask/snaps-sdk';
+import type { JsonRpcRequest, PendingJsonRpcResponse } from '@metamask/utils';
+
+import { setStateHandler, type SetStateParameters, set } from './setState';
+
+describe('snap_setState', () => {
+  describe('setStateHandler', () => {
+    it('has the expected shape', () => {
+      expect(setStateHandler).toMatchObject({
+        methodNames: ['snap_setState'],
+        implementation: expect.any(Function),
+        hookNames: {
+          getSnapState: true,
+          hasPermission: true,
+        },
+      });
+    });
+  });
+
+  describe('implementation', () => {
+    it('sets the encrypted state', async () => {
+      const { implementation } = setStateHandler;
+
+      const getSnapState = jest.fn().mockReturnValue({
+        foo: 'bar',
+      });
+
+      const updateSnapState = jest.fn().mockReturnValue(null);
+      const getUnlockPromise = jest.fn().mockResolvedValue(undefined);
+      const hasPermission = jest.fn().mockReturnValue(true);
+
+      const hooks = {
+        getSnapState,
+        updateSnapState,
+        getUnlockPromise,
+        hasPermission,
+      };
+
+      const engine = new JsonRpcEngine();
+
+      engine.push((request, response, next, end) => {
+        const result = implementation(
+          request as JsonRpcRequest<SetStateParameters>,
+          response as PendingJsonRpcResponse<SetStateResult>,
+          next,
+          end,
+          hooks,
+        );
+
+        result?.catch(end);
+      });
+
+      const response = await engine.handle({
+        jsonrpc: '2.0',
+        id: 1,
+        method: 'snap_setState',
+        params: {
+          key: 'foo',
+          value: 'baz',
+        },
+      });
+
+      expect(getUnlockPromise).toHaveBeenCalled();
+      expect(updateSnapState).toHaveBeenCalledWith({ foo: 'baz' }, true);
```

Comment:
Assert that we are getting the right state too please

Reply:
What do you mean?

Reply:
Assert that `getSnapState` is called with the right param, so we can catch issues like the one I found above 😄

Reply:
Ah, right. Done!